src/app/pages/BorrowPage/components/AddCollateralModal/AddCollateralModal.tsx

```diff
@@ -0,0 +1,336 @@
+  weiToAssetNumberFormat,
+} from 'utils/display-text/format';
+import { LabelValuePair } from 'app/components/LabelValuePair';
+import { TxFeeCalculator } from 'app/pages/MarginTradePage/components/TxFeeCalculator';
+import { ActiveLoan } from 'app/hooks/trading/useGetLoan';
+import { useDenominateAssetAmount } from 'app/hooks/trading/useDenominateAssetAmount';
+import { LoadableValue } from 'app/components/LoadableValue';
+import {
+  BORROW_MAINTENANCE_RATIO,
+  calculateCollateralRatio,
+  calculateLiquidationPrice,
+} from './utils';
+
+type AddCollateralModalProps = {
+  loan: ActiveLoan;
+  onCloseModal: () => void;
+  liquidationPrice?: React.ReactNode;
+  positionSize?: string;
+};
+
+enum Step {
+  FORM,
+  REVIEW,
+}
+
+export const AddCollateralModal: React.FC<AddCollateralModalProps> = ({
+  loan: item,
+}) => {
+  const canInteract = useCanInteract();
+  const tokenDetails = AssetsDictionary.getByTokenContractAddress(
+    item?.collateralToken || '',
+  );
+  const loanToken = AssetsDictionary.getByTokenContractAddress(
+    item?.loanToken || '',
+  );
+  const [amount, setAmount] = useState('');
+  const [step, setStep] = useState(Step.FORM);
+  const { value: balance } = useAssetBalanceOf(tokenDetails.asset);
+  const weiAmount = useWeiAmount(amount);
+
+  const { send, ...tx } = useApproveAndAddMargin(
+    tokenDetails.asset,
+    item.loanId,
+    weiAmount,
+  );
+  const { checkMaintenance, States } = useMaintenance();
+  const topupLocked = checkMaintenance(States.ADD_TO_MARGIN_TRADES);
+
+  const valid = useIsAmountWithinLimits(weiAmount, '1', balance);
+  const { t } = useTranslation();
+
+  const {
+    value: principalAsCollateral,
+    loading: loadingRate,
+  } = useDenominateAssetAmount(
+    loanToken.asset,
+    tokenDetails.asset,
+    item.principal,
+  );
+
+  const currentCollateralRatio = useMemo(
+    () => calculateCollateralRatio(principalAsCollateral, item.collateral),
+    [item.collateral, principalAsCollateral],
+  );
+
+  const currentLiquidationPrice = useMemo(
+    () =>
+      calculateLiquidationPrice(
+        item.principal,
+        item.collateral,
+        BORROW_MAINTENANCE_RATIO,
+      ),
+    [item.collateral, item.principal],
+  );
+
+  const newCollateralAmount = useMemo(
+    () => bignumber(item.collateral).add(weiAmount).toFixed(0),
+    [item.collateral, weiAmount],
+  );
+
+  const newLiquidationPrice = useMemo(
+    () =>
+      calculateLiquidationPrice(
+        item.principal,
+        bignumber(item.collateral).add(weiAmount).toString(),
+        BORROW_MAINTENANCE_RATIO,
+      ),
+    [item.collateral, item.principal, weiAmount],
+  );
+
+  const newCollateralRatio = useMemo(
+    () =>
+      calculateCollateralRatio(
+        principalAsCollateral,
+        bignumber(item.collateral).add(weiAmount).toString(),
+      ),
+    [item.collateral, principalAsCollateral, weiAmount],
+  );
+
+  const handleNextStep = useCallback(() => setStep(Step.REVIEW), []);
+
+  return (
```

Comment:
`render` is way too long, pls extract `Step.FORM` render and `Step.REVIEW` to separate components and use them here, it will help with readability of the code.